src/Moryx.Runtime.Kernel/FileSystem/LocalFileSystem.cs

```diff
@@ -0,0 +1,278 @@
+                await sw.FlushAsync();
+            }
+
+            tree.Hash = await StreamToDiskAsync(stream);
+
+            // Update parent or owner file
+            if (tree.ParentTree == null)
+            {
+                var ownerFile = Path.Combine(_ownerFilesDirectory, tree.FileName);
+                File.WriteAllText(ownerFile, tree.Hash);
+            }
+            else
+            {
+                await WriteTreeAsync(tree.ParentTree);
+            }
+
+            return tree.Hash;
+        }
+
+        private async Task<string> WriteBlobAsync(MoryxFile file, Stream fileStream)
+        {
+            // Create file first
+            var hash = await StreamToDiskAsync(fileStream);
+            file.Hash = hash;
+
+            // Now write the tree recursively 
+            await WriteTreeAsync(file.ParentTree);
+
+            return hash;
+        }
+
+        private async Task<string> StreamToDiskAsync(Stream stream)
+        {
+            var hashPath = HashPath.FromStream(stream);
+
+            // Create directory if necessary
+            var targetPath = hashPath.DirectoryPath(_fsDirectory);
+            try
+            {
+                if (!Directory.Exists(targetPath))
+                    Directory.CreateDirectory(targetPath);
+            }
+            catch (Exception e)
+            {
+                throw LoggedException(e, _logger, _fsDirectory);
+            }
+
+            var fileName = hashPath.FilePath(_fsDirectory);
+            if (File.Exists(fileName))
+                return hashPath.Hash;
+
+            // Write file
+            try
+            {
+                using var fileStream = new FileStream(fileName, FileMode.Create);
+                await stream.CopyToAsync(fileStream);
+                await fileStream.FlushAsync();
+                stream.Position = 0;
+            }
+            catch (Exception e)
+            {
+                throw LoggedException(e, _logger, fileName);
+            }
+
+            return hashPath.Hash;
+        }
+
+        public Stream OpenStream(string hash)
+        {
+            var path = HashPath.FromHash(hash).FilePath(_fsDirectory);
+            return File.Exists(path) ? new FileStream(path, FileMode.Open, FileAccess.Read) : null;
+        }
+
+        public MoryxFileTree ReadTreeByOwner(string ownerKey)
+        {
+            var existingTree = _ownerTrees.FirstOrDefault(ot => ot.FileName == ownerKey);
+            if (existingTree == null)
+            {
+                existingTree = new MoryxFileTree { FileName = ownerKey };
+                _ownerTrees.Add(existingTree);
+            }
+            return existingTree;
+        }
+
+        private MoryxFileTree ReadExtensibleTree(MoryxFileTree tree)
+        {
+            // Read tree from hash
+            var path = HashPath.FromHash(tree.Hash).FilePath(_fsDirectory);
+            if (!File.Exists(path))
+                throw new FileNotFoundException(path);
```

Comment:
since this is also called in the `LoadTrees` method I would expect that the error will be confusing for people using it. I imagine someone booting up the application and for some reason an owner file is missing. I think that could be better explained than this 🙈

Reply:
Doesn't the ReadAllFiles method also throw the exception, when called on a non existing file? 😅 i.e. is there any benefit in this?